server/public/plugin/client.go

```diff
@@ -26,14 +25,69 @@ func ClientMain(pluginImplementation any) {
 	impl.SetAPI(nil)
 	impl.SetDriver(nil)
 
-	pluginMap := map[string]plugin.Plugin{
+	return map[string]plugin.Plugin{
 		"hooks": &hooksPlugin{hooks: pluginImplementation},
 	}
+}
+
+// WithTestContext provides a context typically used to terminate a plugin from a unit test.
+func WithTestContext(ctx context.Context) func(*plugin.ServeConfig) error {
+	return func(config *plugin.ServeConfig) error {
+		if config.Test == nil {
+			config.Test = &plugin.ServeTestConfig{}
+		}
+
+		config.Test.Context = ctx
+
+		return nil
+	}
+}
+
+// WithTestReattachConfigCh configures the channel to receive the ReattachConfig used to reattach
+// an externally launched plugin instance with the Mattermost server.
+func WithTestReattachConfigCh(reattachConfigCh chan<- *plugin.ReattachConfig) func(*plugin.ServeConfig) error {
+	return func(config *plugin.ServeConfig) error {
+		if config.Test == nil {
+			config.Test = &plugin.ServeTestConfig{}
+		}
+
+		config.Test.ReattachConfigCh = reattachConfigCh
+
+		return nil
+	}
+}
+
+// WithTestCloseCh provides a channel that signals when the plugin exits.
+func WithTestCloseCh(closeCh chan<- struct{}) func(*plugin.ServeConfig) error {
+	return func(config *plugin.ServeConfig) error {
+		if config.Test == nil {
+			config.Test = &plugin.ServeTestConfig{}
+		}
 
-	plugin.Serve(&plugin.ServeConfig{
+		config.Test.CloseCh = closeCh
+
+		return nil
+	}
+}
+
+// Starts the serving of a Mattermost plugin over net/rpc. gRPC is not yet supported.
+//
+// Call this when your plugin is ready to start. Options allow configuring plugins for testing
+// scenarios.
+func ClientMain(pluginImplementation any, opts ...func(config *plugin.ServeConfig) error) {
```

Comment:
Just a self-note: This is imported and called on the plugin side. Are there any backwards compatibility concerns with this? I think not but just calling it out

Reply:
Since we version `public` separately, it's opt-in when a plugin upgrades their `go.mod`, but even then the syntax should be backwards compatible thanks to the argument spread. Agree with your conclusion.